### PR TITLE
Switch to GitHub Actions

### PR DESCRIPTION
**Pull request by crazy-max**
Follow-up #208 (cc @thaJeztah)

This PR allows to build credential helpers on GitHub Actions using cross-compilation thanks to [tonistiigi/xx](https://github.com/tonistiigi/xx/) (cc @tonistiigi).

* Containerized environment, no need windows or macos runners
* Switch to GolangCI Lint
* Deb packaging refactored (containerized too)

Run example: https://github.com/crazy-max/docker-credential-helpers/actions/runs/1181534227

@ebriney Is Jenkinsfile still used?

Signed-off-by: CrazyMax <crazy-max@users.noreply.github.com>

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+jobs:
+  test:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        os:
+          - ubuntu-latest
+          - macOS-latest
+          - windows-latest
```
Comment (crazy-max) [inline .github/workflows/build.yml:26]: Tests are also run on specific runners with respect to the target OS and the results are merged and sent to Codecov.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+xx-verify /out/docker-credential-secretservice
+EOT
+
+FROM gobase AS build-darwin
+ARG TARGETOS
+ARG TARGETPLATFORM
+RUN --mount=type=bind,target=. \
+  --mount=type=cache,target=/root/.cache \
+  --mount=type=cache,target=/go/pkg/mod \
+  --mount=type=bind,from=tonistiigi/xx:sdk-extras,src=/xx-sdk,target=/xx-sdk \
```
Comment (crazy-max) [inline Dockerfile:41]: Needs the sdk to build the osxkeychain.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+  --mount=type=bind,from=version,source=/tmp/.ldflags,target=/tmp/.ldflags <<EOT
+set -ex
+mkdir /out
+xx-go install std
```
Comment (crazy-max) [inline Dockerfile:45]: `std` needs to be installed for stdlib.

File: Dockerfile [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+xx-go build -ldflags "$(cat /tmp/.ldflags)" -o /out/docker-credential-wincred.exe ./wincred/cmd/main_windows.go
+xx-verify /out/docker-credential-wincred.exe
+EOT
+
+FROM build-$TARGETOS AS build
+
+FROM scratch AS binaries
+COPY --from=build /out /
+
+FROM debian:bullseye-slim AS build-deb
```
Comment (crazy-max) [inline Dockerfile:68]: Build debian packages for `pass` and `secretservice`

File: docker-bake.hcl [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+target "test" {
+  args = {
+    GO_VERSION = GO_VERSION
+  }
+  target = "test-coverage"
+  output = ["./coverage"]
+  platforms = [
+//    "darwin/amd64",
+//    "darwin/arm64",
+    "linux/amd64",
```
Comment (crazy-max) [inline docker-bake.hcl:51]: Containerized tests are only available for linux platform (that's why a specific job has been added in the build workflow).

File: hack/release [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      tar cvfz "${workdir}/${RELEASE_OUT}/docker-credential-pass-$arch.tar.gz" docker-credential-pass
+      tar cvfz "${workdir}/${RELEASE_OUT}/docker-credential-secretservice-$arch.tar.gz" docker-credential-secretservice
+    )
+  elif [ "$os" = "darwin" ]; then
+    (
+      cd $d
+      set -x
+      #SIGNINGHASH=$(security find-identity -v -p codesigning | grep "Developer ID Application: Docker Inc" | cut -d ' ' -f 4)
+      #xcrun -log codesign -s $(SIGNINGHASH) --force --verbose docker-credential-osxkeychain
+      #xcrun codesign --verify --deep --strict --verbose=2 --display docker-credential-osxkeychain
```
Comment (crazy-max) [inline hack/release:35]: We discussed about that with @tonistiigi to sign our buildx binaries. I'm not sure if we should add secrets on the repo to be able to codesign (can be painful) or if we should create our own self-hosted MacOS GitHub Runner with the identity already available (cc @chris-crone).

File: hack/release [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+      tar cvfz "${workdir}/${RELEASE_OUT}/docker-credential-osxkeychain-$arch.tar.gz" docker-credential-osxkeychain
+    )
+  elif [ "$os" = "windows" ]; then
+    (
+      cd $d
+      set -x
+      #echo ${PFX} | base64 -d > $dockerpfx
+      #signtool sign /fd SHA256 /a /f pfx /p ${PFXPASSWORD} /d Docker /du https://www.docker.com /t http://timestamp.verisign.com/scripts/timestamp.dll docker-credential-wincred.exe
```
Comment (crazy-max) [inline hack/release:43]: Same as previous comment with Windows signtool.

File: secretservice/secretservice_linux_test.go [hunk 8]
```diff
@@ -7,8 +7,14 @@ import (
 	"github.com/docker/docker-credential-helpers/credentials"
 )
 
+// These tests are not runnable from a headless environment such as
+// Docker or a CI pipeline due to the DBus "prompt" interface being called
+// when creating and unlocking a keychain.
```
Comment (crazy-max) [inline secretservice/secretservice_linux_test.go:12]: Need to find a headless CLI tool to be able to run `secretservice` tests suite or we could mock prompter that connects to DBus.

**Comment by ebriney** [comment]
@crazy-max , Jenkinsfile is not used, you can delete it

**Comment by crazy-max** [comment]
@ebriney 

> Jenkinsfile is not used, you can delete it

Done

As @thaJeztah suggested I removed `deb` pkgs creation from the release target. Let me know if something else is missing or need to be done to merge this PR.

**Comment by crazy-max** [comment]
PTAL @thaJeztah @ebriney 

Since https://twitter.com/peter_szilagyi/status/1437646118700175360 I guess it's important to start ditching travis :)

**Comment by crazy-max** [comment]
Removed the GitHub Release step as discussed with @ebriney

**Comment by crazy-max** [comment]
Closing this one as we have moved to GHA in other PRs. Debian packaging will be moved to another repo in a follow-up.

**Pull request closed without merging**